src/main/java/com/feedzai/commons/sql/abstraction/batch/impl/MultithreadedBatch.java

```diff
@@ -0,0 +1,497 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.feedzai.commons.sql.abstraction.batch.impl;
+
+import com.feedzai.commons.sql.abstraction.batch.AbstractPdbBatch;
+import com.feedzai.commons.sql.abstraction.batch.BatchEntry;
+import com.feedzai.commons.sql.abstraction.batch.PdbBatch;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngine;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngineException;
+import com.feedzai.commons.sql.abstraction.engine.configuration.PdbProperties;
+import com.feedzai.commons.sql.abstraction.listeners.BatchListener;
+import com.feedzai.commons.sql.abstraction.listeners.MetricsListener;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Marker;
+import org.slf4j.MarkerFactory;
+
+import javax.inject.Inject;
+import java.time.Duration;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * A Batch that periodically flushes pending insertions to the database using multiple threads/connections.
+ *
+ * @author José Fidalgo (jose.fidalgo@feedzai.com)
+ */
+public class MultithreadedBatch extends AbstractPdbBatch implements PdbBatch {
+
+    /**
+     * The logger.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(MultithreadedBatch.class);
+
+    /**
+     * The confidential logger.
+     */
+    private final Logger confidentialLogger;
+
+    /**
+     * The dev Marker.
+     */
+    private static final Marker DEV = MarkerFactory.getMarker("DEV");
+
+    /**
+     * A delay to make sure that when the next periodic flush is triggered, it occurs after the batch timeout period
+     * has passed since the last flush occurred (unless the flush was triggered by batch size).
+     */
+    private static final int SALT = 100;
```

Comment:
I just kept it as it was before, but I don't think it's useless.
I had to think about its purpose, and my understanding is expressed in the javadoc.
I think this exists to try to force the periodic flush to always occur after "batchTimeout" has passed since the last flush;
this has been here since the repository was created, so I assume it was added because sometimes the periodic task was triggered before `now - lastFlush` (so the task would do nothing until the next flush)